Multichain

xvi10 opened the pull request.

dmax10 committed:
Add multichain handler contract

dmax10 committed:
Add multichain withdrawal

dmax10 committed:
Add multichain provider interface, multichain provider signature contract and multichain provider utils lib

dmax10 committed:
Add multichain errors, small refactor

dmax10 committed:
Simplify LayerZeroProvider contract (keep only lzCompose for now)

dmax10 committed:
Fix compilers error, add source chain balance key under the allowed base keys

dmax10 committed:
Remove multichain controller role, use controller role for provider contract

dmax10 committed:
Add deploy scripts

dmax10 committed:
Add multichain utils

dmax10 committed:
Store open position price impact on position open instead of charging it directly, charge the total price impact (open + close) on position close

This allows total price impact to be capped, with the excess being claimable

dmax10 committed:
Keep the original event key since it is used for e.g. analytics

dmax10 committed:
Remove comments

dmax10 committed:
Track impact pending amount using a hash from account and market.

Using the positionKey which is the hash of account, market, collateralToken, and isPositive would make it harder to track the impact for the same position on increase/decrease.

dmax10 committed:
Remove combined key logic for storing the pending impact

dmax10 committed:
Use position store instead of the data store for storing/retrieving position impact pending amount

dmax10 committed:
Remove blank lines

dmax10 committed:
Use balanceWasImproved instead of priceImpactUsd > 0 as forPositiveImpact

dmax10 committed:
Use balanceWasImproved for crossover rebalance

dmax10 committed:
Add token address to OracleTimestampsAreSmallerThanRequired error

dmax10 committed:
Update affiliate reward tests when balance was improved

dmax10 committed:
Update comments

dmax10 committed:
Revert "Add token address to OracleTimestampsAreSmallerThanRequired error"

This reverts commit 77832ffd1735815d551b896255a389efef7c2e2b.

dmax10 committed:
Order.market should be validated for swap orders

dmax10 committed:
Revert "Order.market should be validated for swap orders"

This reverts commit a09709a98ce804c6d557f609908a89f9c382882d.

dmax10 committed:
Add support for bytes type in DataStore (bytesValues, getBytes, setBytes, removeBytes)

dmax10 committed:
WIP add data field for deposits, orders, withdrawals, shifts etc

dmax10 committed:
WIP Allow setting a CLAIMABLE_COLLATERAL_REDUCTION_FACTOR for (account + time + market) key for price impact rebates

dmax10 committed:
Move claimable reduction method from ConfigSyncer to Config

dmax10 committed:
Add ConfigUtils lib and move logic from the Config to reduce the contract size

dmax10 committed:
Revert changes for the DataStore

dmax10 committed:
Replace bytes data field with bytes32[] dataList

dmax10 committed:
Fix tests: default dataList to empty array

dmax10 committed:
Fix tests: increase expectedPropsLength by one for the structs where the dataList field was added

dmax10 committed:
Remove IConfigUtils interface and import ConfigUtils directly.

Since the lib methods are external, Config contract size is still reduced. No need to use an interface.

dmax10 committed:
Remove contructor param from Config

dmax10 committed:
Add data length validation

dmax10 committed:
Add dataList to Orders

dmax10 committed:
Revert "Refactor ExchangeRouter to reduce it's size"

This reverts commit 2bc7efab8d0f6046c0f55050a89e7b952200bb41.

dmax10 committed:
Move setClaimableCollateralReductionFactorForAccount logic from Config to ConfigUtils

dmax10 committed:
Comments update

dmax10 committed:
Revert dataList for GlvShift

GlvShift doesn't need dataList since createGlvShift is only used internally by keepers, so integrations can't use it

dmax10 committed:
Remove dataList from ADL orders

xdev10 committed:
Merge branch 'gasless' into multichain

# Conflicts:
#	contracts/error/Errors.sol

xdev10 committed:
Merge branch 'v2.2-branch' into multichain

xdev10 committed:
update yarn

dmax10 committed:
WIP multichain

dmax10 committed:
WIP multichain

dmax10 committed:
Refactor GlvDepositUtils to reduce bytecode size

Move _validateFirstGlvDeposit and _getMintAmount internal methods from GlvDepositUtils to the newly created GlvDepositHelper library.

xvi10 commented on contracts/multichain/MultichainRouter.sol:
```diff
@@ -0,0 +1,164 @@
+        // TODO: confirm virtual account is used correctly
+        address virtualAccount = MultichainUtils.getVirtualAccount(account, sourceChainId);
+
+        return _createDeposit(relayParams.tokenPermits, relayParams.fee, params, virtualAccount);
```
should use _createDeposit(..., account, chainId) here instead of virtualAccount

the GM tokens should be sent to MultichainVault and the account's multichain GM balance should be increased

so would need a deposit.chainId as well

in execute deposit, if the deposit.chainId is "0" this means that the deposit would be a normal deposit, so the GM tokens would be sent directly to the user, if the deposit.chainId is not zero, then the user's multichain GM balance should be increased

dmax10 replied:
Updated. Named it srcChainId though.

xvi10 commented on contracts/multichain/MultichainRouter.sol:
```diff
@@ -0,0 +1,164 @@
+            eventEmitter: eventEmitter,
+            // TODO: confirm Contracts struct can be modified --> replace `OrderVault orderVault;` field with `StrictBank vault;`
+            // otherwise, should probably overridde _handleRelay
+            orderVault: OrderVault(payable(multichainVault))
```
@gdev8317 is it ok for Max to change Contracts { OrderVault orderVault } into Contracts { StrictBank bank }?

dmax10 replied:
Changed it.

xvi10 commented on contracts/multichain/MultichainRouter.sol:
```diff
@@ -0,0 +1,164 @@
+
+        // TODO: confirm where funds are bridged. Are they sent to this contract, to MultichainProvider, or directly to MultichainVault?
+        // if funds are sent directly to MultichainVault, _sendTokens would not be needed, but then where is the relay fee being paid from?
+        _sendTokens(
```
funds should be bridged to MultichainVault directly

> if funds are sent directly to MultichainVault, _sendTokens would not be needed

I don't quite understand this, to create a deposit, the tokens would still need to be sent from MultichainVault to DepositVault?

> then where is the relay fee being paid from?

relay fee should be sent from MultichainVault, from the user's multichain balance

residual fees should be sent to the deposit vault

dmax10 replied:
> I don't quite understand this, to create a deposit, the tokens would still need to be sent from MultichainVault to DepositVault?

I thought so. DepositUtils.createDeposit does a recordTransferIn, which assumes tokens have already been transferred to DepositVault

xvi10 commented on contracts/multichain/MultichainUtils.sol:
```diff
@@ -7,6 +7,6 @@ pragma solidity ^0.8.0;
  */
 library MultichainUtils {
     function getVirtualAccount(address account, uint256 sourceChainId) internal pure returns (address) {
```
should remove this function

dmax10 replied:
removed

dmax10 committed:
Rename glv deposit utility lib

dmax10 committed:
Merge branch 'gasless' into multichain

dmax10 committed:
Rename deploy script

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Make BaseGelatoRelayRouter._sendTokens virtual to be able to override it in MultichainRouter

dmax10 committed:
fix compile errors

dmax10 committed:
Rename MultichainHandler in MultichainVaultHandler

dmax10 committed:
Rename MultichainHandler in MultichainVaultHandler

dmax10 committed:
Add GaslessCreateDepositParams

dmax10 committed:
Add chainId to Deposit.Props

dmax10 committed:
Add chainId to Deposit.Props

dmax10 committed:
Update multichain createDeposit logic

dmax10 committed:
Update multichain createDeposit logic

dmax10 committed:
Emit dataList for createDeposit, createWithdrawal, createOrder, updateOrder, createShift, createGlvDeposit, createGlvWithdrawal

dmax10 committed:
Emit dataList for createDeposit, createWithdrawal, createOrder, updateOrder, createShift, createGlvDeposit, createGlvWithdrawal

dmax10 committed:
Rename sourceChain* to multichain*

dmax10 committed:
Rename sourceChain* to multichain*

dmax10 committed:
Add multichain increase and decrease utils methods

dmax10 committed:
Add multichain increase and decrease utils methods

dmax10 committed:
Add chainId to Withdrawal, Shift, GlvDeposit, GlvWithdrawal

gdev8317 commented on contracts/multichain/MultichainRouter.sol:
```diff
@@ -0,0 +1,183 @@
+import "./MultichainUtils.sol";
+
+contract MultichainRouter is GelatoRelayRouter {
+    struct GaslessCreateDepositParams {
```
i would suggest to call it MulticallCreateDepositParams (and use multichain instead of gasless in otehr names) because GelatoRelayRouter is also gasless

gdev8317 replied:
and maybe chainId, longTokenAmount and shortTokenAmount can be passed as a separate arguments and remove the extra nesting? just thinking, at first glance it could simplify the contract interface

dmax10 replied:
Changed to:
```
struct MultichainCreateDepositParams {
        uint256 desChainId;
        uint256 longTokenAmount;
        uint256 shortTokenAmount;
        DepositUtils.CreateDepositParams createDepositParams;
    }
```
and replaced gasless with multichain as well.

dmax10 replied:
Didn't change the nesting yet, but could do that as well if needed.

gdev8317 commented on contracts/multichain/MultichainRouter.sol:
```diff
@@ -0,0 +1,183 @@
+            tokenPermits,
+            fee,
+            account,
+            NonceUtils.getKey(contracts.dataStore, NonceUtils.getCurrentNonce(dataStore) + 1), // calculate next key without incrementing
```
would it be the same as `NonceUtils.getNextKey(dataStore)`?

dmax10 replied:
Yes, same logic, the difference is that getNextKey also writes the incremented value (it is named "get" but it does "set" as well).

gdev8317 replied:
oh right, not ideal naming, would be good to change it sometime

gdev8317 commented on contracts/multichain/MultichainUtils.sol:
```diff
@@ -2,11 +2,18 @@
-    function getVirtualAccount(address account, uint256 sourceChainId) internal pure returns (address) {
-        return address(uint160(uint256(keccak256(abi.encode("GMX Multichain", account, sourceChainId)))));
+    function increaseBalance(DataStore dataStore, uint256 chainId, address account, address token, uint256 amount) internal {
+        dataStore.decrementUint(Keys.multichainBalanceKey(chainId, account, token), amount);
```
i guess it should be incrementUint instead of decrementUint

dmax10 replied:
Changed it.

gdev8317 commented on contracts/multichain/MultichainRouter.sol:
```diff
@@ -0,0 +1,183 @@
+        Contracts memory contracts = Contracts({
+            dataStore: dataStore,
+            eventEmitter: eventEmitter,
+            // TODO: confirm Contracts struct can be modified --> replace `OrderVault orderVault;` field with `StrictBank vault;`
```
i think it's okay to change type to StrictBank, yes

dmax10 replied:
Changed it.

gdev8317 commented on contracts/multichain/MultichainEventUtils.sol:
```diff
@@ -16,56 +16,56 @@ library MultichainEventUtils {
+        eventData.addressItems.setItem(1, "account", account);
 
         eventData.uintItems.initItems(2);
         eventData.uintItems.setItem(0, "amount", amount);
```
wondering it if makes sense to only pass deposit key to this event instead of all other fields

events consume quite a lot of gas, unless i'm missing something account/token/amount could be retrieved from corresponding DepositCreated event later

but tbh i'm not sure if the difference worth it. events just consume a lot of gas already unfortunately

dmax10 committed:
Fix increaseBalance: incrementUint instead of decrementUint

dmax10 committed:
Set chainId for Deposit on executeShift, remove position chainId related comments

dmax10 committed:
Set chainId for Deposit on executeShift, remove position chainId related comments

dmax10 committed:
WIP multichain deposits and gm logic

dmax10 committed:
WIP multichain deposits and gm logic

dmax10 committed:
Update multichain logic for fee payment

dmax10 committed:
Update multichain logic for fee payment

dmax10 committed:
Remove chainId from multichain balance

dmax10 committed:
Remove chainId from multichain balance

dmax10 committed:
Rename chainId to srcChainId

dmax10 committed:
Move srcChainId into RelayParams

dmax10 committed:
Emit srcChainId for createWithdrawal, createShift, createGlvDeposit, createGlvWithdrawal

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Fix tests after merging gasless branch

dmax10 committed:
Use srcChainId instead of block.chainid to _validateCall

dmax10 committed:
Use block.chainid as the desChainId to genearte gaslessCreateDepositParamsStructHash

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Add in and out multichain transfers methods to MultichainUtils lib. Refactor all related contracts.

dmax10 committed:
Add in and out multichain transfers methods to MultichainUtils lib. Refactor all related contracts.

dmax10 committed:
Add multichainVault address to contracts and stucts

dmax10 committed:
Validate desChainId provided is block.chainid, rename gasless to multichain

dmax10 committed:
Add srcChainId to _sendTokens

dmax10 committed:
Change Contracts { OrderVault orderVault } into Contracts { StrictBank bank }

dmax10 committed:
Split CreateDepositParams into subgroups

dmax10 committed:
Split CreateDepositParams into subgroups

dmax10 committed:
Add srcChainId to orders

dmax10 committed:
Add srcChainId to orders

dmax10 committed:
Add srcChainId to _sendTokens

dmax10 committed:
Rename multichainId to srcChainId, remove second topic log

dmax10 committed:
Rename multichainId to srcChainId, remove second topic log

dmax10 committed:
reorder method params

dmax10 committed:
Record residualFee under user's multichain balance when the residualFeeReceiver is the multichainVault, remove comments

dmax10 committed:
Add srcChiainId field to CreateOrderParams struct

dmax10 committed:
Add srcChiainId field to CreateOrderParams struct

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
gasless merge fixes

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Implement multichain createWithdrawal

Commit:
pr fix

Commit:
pr fix

dmax10 committed:
Use multichainVault as withdrawal receiver for GM/GLV for source chain transfers, record output tokens under user's multichain balance

dmax10 committed:
Implement multichain createGlvDeposit

dmax10 committed:
Use MultichainUtils as external lib to avoid increasing the size of the contracts using it

dmax10 committed:
Implement multichain createGlvWithdrawal

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Fix function params order

dmax10 committed:
Implement multichain createShift

dmax10 committed:
Fix refund fee receiver for multichain actions

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

gmxer committed:
set default maxPnlFactorForTraders for VIRTUAL and PENGU markets

dmax10 committed:
Refactor BaseGelatoRelayRouter, GelatoRelayRouter, MultichainRouter to outsource relay related constants and functions into an external RelayUtils lib

dmax10 committed:
Refactor MultichainRouter by spliting it's logic into MultichainGmRouter and MultichainGlvRouter to reduce contract size

dmax10 committed:
Add TransferRequest[] param to create* multichain actions and _sendTokens by looping on this array. Refactor all related typehashes and helper functions.

dmax10 committed:
Add TransferRequest[] param to create* multichain actions and _sendTokens by looping on this array. Refactor all related typehashes and helper functions.

gmxer committed:
upd maxPnlFactorForTraders for new markets

dmax10 committed:
Refactor GlvDepositUtils to move the "execute" part into ExecuteGlvDepositUtils to reduce contract size

dmax10 committed:
Uncomment relay fee refund logic after solving the contract size error

dmax10 committed:
Refactor to add addresses subgroup to CreateGlvDepositParams to fix typehash slot too deep error

dmax10 committed:
Refactor to add addresses subgroup to CreateGlvWithdrawalParams to fix typehash slot too deep error

dmax10 committed:
Remove "immutable" from state variables to reduce contract size bellow the deploying limit

dmax10 committed:
Remove srcChainId from RelayParams struct

dmax10 committed:
Remove outdated contracts

dmax10 committed:
Remove outdated contracts

dmax10 committed:
Remove srcChainId from RelayParams struct (continued)

dmax10 committed:
Move desChainId from multichain structs into RelayParams struct, remove multichain struct from create functions, remove top level multichain typehases

dmax10 committed:
Add srcChainId to depositHandler.createDeposit, small DepositUtils refactor to solve stack too deep error

dmax10 committed:
Add srcChainId to depositHandler.createDeposit, small DepositUtils refactor to solve stack too deep error

dmax10 committed:
Add srcChainId to withdrawalHandler.createWithdrawal, small WithdrawalUtils refactor to solve stack too deep error

dmax10 committed:
Add srcChainId to withdrawalHandler.createWithdrawal, small WithdrawalUtils refactor to solve stack too deep error

dmax10 committed:
Add srcChainId to shiftHandler.createShift

dmax10 committed:
Add srcChainId to shiftHandler.createShift

dmax10 committed:
Add srcChainId to glvHandler (createGlvDeposit, createGlvWithdrawal), small GlvWithdrawalUtils refactor to solve stack too deep error

dmax10 committed:
Add srcChainId to orderHandler.createOrder, small OrderUtils refactor to solve stack too deep error

dmax10 committed:
Add srcChainId to orderHandler.createOrder, small OrderUtils refactor to solve stack too deep error

dmax10 committed:
Remove srcChainId from non-multichain orders

dmax10 committed:
Add MultichainGmRouter: createOrder, updateOrder, cancelOrder

dmax10 committed:
Rename contract, remove comment, add modifier

dmax10 committed:
Rename contract, remove comment, add modifier

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
gassless merge fix

dmax10 committed:
Refactor: add function to validate destination chain

dmax10 committed:
Refactor: add function to validate destination chain

dmax10 committed:
Add bridgeOut function, validate multichain provider is whitelisted, emit bridge in/out events, add bridgeOut thypehashes

dmax10 committed:
wip provider bridgeOut

dmax10 committed:
Merge branch 'gasless' into multichain

dmax10 committed:
Fix gasless merge

dmax10 committed:
Fix gasless merge

dmax10 committed:
Add dataList field to relay create order params

dmax10 committed:
Fix signatures and missing fields for gasless tests

dmax10 committed:
Include desChainId into relay signature

dmax10 committed:
Validate desChainId against block.chainid (for same chain actions, chainId and desChainId are the same)

dmax10 committed:
Validate desChainId against block.chainid (for same chain actions, chainId and desChainId are the same)

dmax10 committed:
Include dataList into relay signature, remove srcChainId from multichain typehashes

dmax10 committed:
Add deploy script for MultichainGmRouter

dmax10 committed:
Fix 'from' param for transferOut

dmax10 committed:
Fix multichain createDeposit typehash (without transferRequests)

dmax10 committed:
WIP test multichain createDeposit flow

dmax10 committed:
change token permit nonce and srcChainId

dmax10 committed:
Add provider specific function param on bridgeOut

dmax10 committed:
Fix multichain transferOut logic

dmax10 committed:
Remove double validation for desChainId

dmax10 committed:
Remove double validation for desChainId

dmax10 committed:
Change transferRequests structure, update typehashes

dmax10 committed:
Change transferRequests structure, update typehashes

dmax10 committed:
Test multichain deposit

dmax10 committed:
Fix ETH deposit and fee payment

dmax10 committed:
Fix ETH deposit and fee payment

dmax10 committed:
Split CreateWithdrawalParams into subgroups to solve relay typehash error

dmax10 committed:
Add error params

dmax10 committed:
WIP test multichain withdrawal

Commit:
add timelockconfigcontroller

Commit:
squash - refactor callbacks to eventdata

Commit:
squash - refactor callbacks to eventdata

Commit:
squash - refactor callbacks to eventdata

xvi10 committed:
Merge pull request #115 from gmx-io/refactor-callbacks-to-event-structs

refactor callbacks to eventdata

Commit:
Merge branch 'refactor-callbacks-to-event-structs' of github.com:gmx-io/gmx-synthetics into refactor-callbacks-to-event-structs

Commit:
fix tests

Commit:
Merge branch 'refactor-callbacks-to-event-structs' into multichain

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Fix tests after gasless merge

dmax10 committed:
test multichain withdrawal

Commit:
timelock controller

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Revert "immutable" to state variable.

It was removed as a temporary fix to reduce the contract size and it's not needed anymore. Different refactoring was implemented

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Split CreateShiftParams into subgroups to solve relay typehash error

dmax10 committed:
Add bridgeIn and move bridgeOut from MultichainOrderRouter into MultichainTransferRouter

dmax10 committed:
Test multichain withdrawal (wip)

dmax10 committed:
Test multichain glv deposit, fix typehases, add deploy script

Commit:
rebase

dmax10 committed:
Test multichain glv withdrawal

dmax10 committed:
Fix refund fee. Skip updating user's multichain balance if the refund amount is zero

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Fix params order for gasless _createOrder

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Add multichain support to orders

dmax10 committed:
Add script to deployMultichainOrderRouter

dmax10 committed:
Fix fee refund error to non initialised multichain vault

dmax10 committed:
Test create multichain order

dmax10 committed:
Test update multichain order

dmax10 committed:
Test cancel multichain order

dmax10 committed:
Add back "immutable" which was temporarily added to reduce contract size, add gasless validation

Commit:
fix tests

dmax10 committed:
Inherit BaseGelatoRelayRouter instead of GelatoRelayRouter into MultichainRouter

dmax10 committed:
Handle partial fee payment for multichain update/cancel orders

dmax10 committed:
Inherit BaseRouter into MultichainTransferRouter for multicall functionality

It's actually inherited into BaseGelatoRelayRouter instead of MultichainTransferRouter due to inheritance structure.
Since SubaccountGelatoRelayRouter was getting too big, some of it's logic is moved into a new external lib SubaccountRelayUtils

dmax10 committed:
Set HOLDING_ADDRESS as the refundReceiverAddress if position collateral was used to pay the execution fee, small refactor to solve stack too deep error in OrderUtils

dmax10 committed:
Emit PositionCollateralUsedForExecutionFee event

dmax10 committed:
Update multichain events, validate gasless feature for all multichain actions

dmax10 committed:
Check keeper fee on multichain deposit

dmax10 committed:
Install @stargatefinance/stg-evm-v2

stargatefinance/stg-evm-v2 requires node  version ">=20" and @poanet/solidity-flattener requires node version ">=16 <=18".

Remove poanet/solidity-flattener since hardhat toolbox already has the flattener included.

dmax10 committed:
Enable multichain bridging out, add deploy script and test setup, other small refactoring

dmax10 committed:
Rename multichain test file

dmax10 committed:
auto format

Commit:
update scripts

Commit:
Merge branch 'multichain' into timelock-refactor

Commit:
update CI node version

xvi10 committed:
Merge pull request #117 from gmx-io/timelock-refactor

Timelock refactor

dmax10 committed:
Fix multichain shift (the withdawal part should go into the shiftVault instead of the multicahinVault)

dmax10 committed:
Merge branch 'v2.2-branch' into multichain

dmax10 committed:
Fix tests after merge
